docs/source/checkpoints/dist_ckpt.rst

```diff
@@ -0,0 +1,392 @@
+Distributed checkpoints
+=======================
+
+This guide provides details about the distributed checkpoints format from Megatron-Core.
+
+
+Introduction
+------------
+
+Model parallel training requires parallelism-aware checkpointing.
+Megatron-Core provides a checkpointing library capable of handling all types of parallelisms used in LLMs training.
+Although distributed checkpointing library is targeted at Megatron-Core model, it can be used with other models as well, provided an appropriate integration.
+
+The library provides two main entrypoints: `dist_checkpointing.save` and `dist_checkpointing.load` which are meant to replace the `torch.save` and `torch.load` in the regular checkpointing flow.
+Apart from that it provides mechanism to define different types of local tensors placement in the global checkpoint.
+
+
+Basic sharding
+--------------
+
+The main way to define relationship of a plain local PyTorch tensor to tensors on other ranks is by wrapping it in a `ShardedTensor` class.
+This allows to express the fact that a given local tensor is part of a larger *grid* of tensors of a given shape at a given offset.
+Instead of saving a simple state dict with `torch.Tensor`s, we save a *sharded* state dict with `dist_checkpointing.ShardedTensor`s.
+
+Example: assume we have a 128 elements tensor divided equally across the whole workload which we want to save and load with different number of ranks.
+
+.. code-block:: python
+    from pathlib import Path
+
+    import torch
+
+    from megatron.core import dist_checkpointing
+
+    # Setup
+    ckpt_root = Path('/tmp/checkpoints')
+    native_ckpt_root = ckpt_root / 'native'
+    native_ckpt_root.mkdir(exist_ok=True, parents=True)
+    dist_ckpt_root = ckpt_root / 'dist_ckpt'
+    dist_ckpt_root.mkdir(exist_ok=True, parents=True)
+
+    torch.distributed.init_process_group()
+    world_size = torch.distributed.get_world_size()
+    rank = torch.distributed.get_rank()
+
+    # Local tensor to save
+    assert 128 % world_size == 0
+    num_elems_per_rank = 128 // world_size
+    local_ten = torch.arange(start=num_elems_per_rank * rank,
+                             end=num_elems_per_rank * (rank + 1))
+
+    # Native checkpoint save
+    state_dict = {
+        'weight': local_ten
+    }
+    torch.save(state_dict, native_ckpt_root / f'ckpt_{rank}.pt')
+
+    # Distributed checkpoint save
+    sharded_state_dict = {
+        'weight': dist_checkpointing.ShardedTensor.from_rank_offsets('weight', local_ten, (0, rank, world_size))
+    }
+    dist_checkpointing.save(sharded_state_dict, dist_ckpt_root)
+
+During load the distributed checkpoint can be easily read even if the job size changes (contrary to native checkpoints that require the same number of ranks).
```

Comment:
During load, the distributed checkpoint can be easily read even if the job size changes (contrary to native checkpoints that require the same number of ranks).